### PR TITLE
Adding securityContext

### PR DESCRIPTION
**Pull request by jtyr**
This PR is adding `securityContext` to make the Jobs created by this controller more secure. The default `securityContext` is set to comply with the `Restricted` [Pod Security Standard](https://kubernetes.io/docs/concepts/security/pod-security-standards/) profile but it can be further tweaked via the `podSecurityContext` and the `securityContext`  fields of the `HelmChart` Custom Resource. Jobs created by the controller are further hardened by enabling `readOnlyRootFilesystem` and enabling write access only to `tmpfs` mounts `/home/klipper-helm/.(helm|cache|config)`.

### CODE REVIEW

File: pkg/apis/helm.cattle.io/v1/types.go [hunk 1]
```diff
@@ -19,23 +19,25 @@ type HelmChart struct {
...
+	HelmVersion        string                        `json:"helmVersion,omitempty"`
+	Bootstrap          bool                          `json:"bootstrap,omitempty"`
+	ChartContent       string                        `json:"chartContent,omitempty"`
+	JobImage           string                        `json:"jobImage,omitempty"`
+	BackOffLimit       *int32                        `json:"backOffLimit,omitempty"`
+	Timeout            *metav1.Duration              `json:"timeout,omitempty"`
+	FailurePolicy      string                        `json:"failurePolicy,omitempty"`
+	AuthSecret         *corev1.LocalObjectReference  `json:"authSecret,omitempty"`
+	PodSecurityContext *corev1.PodSecurityContext    `json:"podSecurityContext,omitempty"`
+	SecurityContext    *corev1.SecurityContext       `json:"securityContext,omitempty"`
```
Comment (brandond) [inline pkg/apis/helm.cattle.io/v1/types.go:40]: Can you put these under the empty line below, to avoid respacing all the previous fields? That's the point of the empty line before the fields with longer names.

File: pkg/controllers/chart/chart.go [hunk 2]
```diff
@@ -51,11 +52,26 @@ const (
...
-	deletePolicy         = metav1.DeletePropagationForeground
-	DefaultJobImage      = "rancher/klipper-helm:v0.8.2-build20230815"
-	DefaultFailurePolicy = FailurePolicyReinstall
-	defaultBackOffLimit  = pointer.Int32(1000)
+	commaRE                   = regexp.MustCompile(`\\*,`)
+	deletePolicy              = metav1.DeletePropagationForeground
+	DefaultJobImage           = "rancher/klipper-helm:v0.8.2-build20230815"
+	DefaultFailurePolicy      = FailurePolicyReinstall
+	defaultBackOffLimit       = pointer.Int32(1000)
+	defaultPodSecurityContext = &corev1.PodSecurityContext{
```
Comment (brandond) [inline pkg/controllers/chart/chart.go:60]: Same request - put an empty line above these to avoid unnecessary whitespace changes on the previous lines.

File: pkg/controllers/chart/chart.go [hunk 3]
```diff
@@ -831,3 +893,66 @@ func hashObjects(job *batch.Job, objs ...metav1.Object) {
 func setBackOffLimit(job *batch.Job, backOffLimit *int32) {
 	job.Spec.BackoffLimit = backOffLimit
 }
+
+func deepCopyStruct(source interface{}) interface{} {
```
Comment (brandond) [inline pkg/controllers/chart/chart.go:897]: Why do we need custom deepcopy functions here? These are provided by Kubernetes for all the core types. See for example:
* https://pkg.go.dev/k8s.io/api/core/v1#SecurityContext.DeepCopy
* https://pkg.go.dev/k8s.io/api/core/v1#PodSecurityContext.DeepCopy

File: pkg/controllers/chart/chart.go [hunk 4]
```diff
@@ -377,6 +393,9 @@ func job(chart *v1.HelmChart, apiServerPort string) (*batch.Job, *corev1.Secret,
 		chartName = chart.Name + "/" + chart.Spec.Chart
 	}
 
+	podSecurityContext := deepCopyStruct(defaultPodSecurityContext).(*corev1.PodSecurityContext)
+	securityContext := deepCopyStruct(defaultSecurityContext).(*corev1.SecurityContext)
```
Comment (brandond) [inline pkg/controllers/chart/chart.go:397]: Do we really need to set both? As the upstream docs note:
> PodSecurityContext holds pod-level security attributes and common container settings. Some fields are also present in container.securityContext. Field values of container.securityContext take precedence over field values of PodSecurityContext.

Is there something specific we're setting in the SecurityContext, that we couldn't get just by setting the PodSecurityContext?

**Review by brandond — CHANGES_REQUESTED** [review]
nits/questions

**Review by jtyr — COMMENTED** [review]
Thanks for the review. All issues should be fixed now.

File: pkg/controllers/chart/chart.go [hunk 5]
```diff
@@ -831,3 +894,38 @@ func hashObjects(job *batch.Job, objs ...metav1.Object) {
...
+					targetField.Set(sourceField)
+				}
+			}
+		}
+	}
+}
+
+func setSecurityContext(job *batch.Job, chart *v1.HelmChart) {
+	if chart.Spec.PodSecurityContext != nil {
+		deepMergeStruct(job.Spec.Template.Spec.SecurityContext, chart.Spec.PodSecurityContext)
```
Comment (brandond) [inline pkg/controllers/chart/chart.go:925]: I'm still not understanding why this is necessary. If chart.Spec.PodSecurityContext/SecurityContext is set, shouldn't we just set that value in the chart, and skip all this deep copying and merging? I've not seen this sort of behavior as necessary in any of our other controllers that allow setting these fields from a higher level spec.

**Review by brandond — CHANGES_REQUESTED** [review]

File: pkg/controllers/chart/chart.go (hunk 5 above)
Comment (jtyr) [reply pkg/controllers/chart/chart.go:925]: The deep merge is necessary because the `podSecurityContext` and `securityContext` are both maps. If the user would like to change just a single field, he/she would have to define all the default fields plus his/her changed field. With the deep merge, the defaults are _updated_ with the user-specified fields recursively. We could remove the deep merge, but I think it's more user friendly if we keep it. I guess, you didn't need any deep copy in other controllers because those controllers don't use maps but only primitive types (string, bool, ...)?

Comment (brandond) [reply pkg/controllers/chart/chart.go:925]: > The deep merge is necessary because the podSecurityContext and securityContext are both maps.

No they're not? They're structs.

https://pkg.go.dev/k8s.io/api/core/v1#SecurityContext
https://pkg.go.dev/k8s.io/api/core/v1#ContainerSecurityContext

> If the user would like to change just a single field, he/she would have to define all the default fields plus his/her changed field.

Yes. The expectation is generally that the field from the spec **sets** the field in the job, not that it is merged with the defaults.  See for example:
https://github.com/rancher/system-upgrade-controller/blob/af373602d649f108321eb4c784d490972499ad9f/pkg/upgrade/container/container.go#L34-L38

Comment (jtyr) [reply pkg/controllers/chart/chart.go:925]: > No they're not? They're structs.

In the YAML terminology it's called a [mapping or `map`](https://yaml.org/spec/1.2.2/#mapping).

> Yes. The expectation is generally that the field from the spec **sets** the field in the job, not that it is merged with the defaults. If the user wants to extend the defaults, they must explicitly include them. See for example: https://github.com/rancher/system-upgrade-controller/blob/af373602d649f108321eb4c784d490972499ad9f/pkg/upgrade/container/container.go#L34-L38

I understand. The problem of having to set the `securityContext` always as fully defined map is that if the user wants to change just a single field and keep all the other fields as they are set by default, the user must copy&paste _all_ the fields instead of just setting the fields of interest. The deep merge principle aligns with how Helm charts work - modification to a single field in a map _updates_ the default map values set in the `values.yaml` file.

I can remove the deep merge if would like to maintain compatibility with your other controllers. Please just let me know.

Comment (brandond) [reply pkg/controllers/chart/chart.go:925]: yes, but we're not dealing with yaml here - we're dealing with golang structs.

The expected behavior for our controllers is that fields set in the spec override defaults, instead of merging with them. That's what I would look for here.

Comment (jtyr) [reply pkg/controllers/chart/chart.go]: It should be fixed in the last commit.

**Review by brandond — APPROVED** [review]

**Review by brandond — APPROVED** [review]
Can you also call out the other changes in the PR description? Namley, that in addition to making the SecurityContexts configurable, it also hardens the **default** values by:
* Enabling ReadOnlyRootFilesystem
* Enabling RunAsNonRoot
* Dropping all capabilities
* Adding tmpfs mounts for `/home/klipper-helm/.(helm|cacheconfig)`

**Comment by brandond** [comment]
Looks like one of the tests is failing:
```
  [FAILED] Expected
      <v1.PodSecurityContext>: {
          SELinuxOptions: nil,
          WindowsOptions: nil,
          RunAsUser: nil,
          RunAsGroup: nil,
          RunAsNonRoot: false,
          SupplementalGroups: nil,
          FSGroup: nil,
          Sysctls: nil,
          FSGroupChangePolicy: nil,
          SeccompProfile: nil,
      }
  to equal
      <v1.PodSecurityContext>: {
          SELinuxOptions: nil,
          WindowsOptions: nil,
          RunAsUser: nil,
          RunAsGroup: nil,
          RunAsNonRoot: false,
          SupplementalGroups: nil,
          FSGroup: nil,
          Sysctls: nil,
          FSGroupChangePolicy: nil,
          SeccompProfile: {
              Type: "RuntimeDefault",
              LocalhostProfile: nil,
          },
      }
```

**Review by brandond — APPROVED** [review]

**Review by manuelbuil — APPROVED** [review]
